BQSR on spark

droazen opened the pull request:
-Ported BQSR and ApplyBQSR to spark.

-Refactored the BQSR engine so that all versions of BQSR (walker, dataflow,
 and spark) call into a common engine, and removed duplicated versions
 of the engine from the codebase.

droazen commented:
For @davidadamsphd to review, please.

If @jean-philippe-martin could also review the modifications to the dataflow BQSR, it would be much appreciated (keeping in mind that we still need to rebase on top of the optimized BQSR branch).

davidadamsphd commented:
Sooooo, this is a huge PR. Can you please give some hints on how best to review (where there many find-replaces that I can just skim over, new files that are really carbon copies)?

laserson commented on src/main/java/org/broadinstitute/hellbender/tools/spark/BaseRecalibratorSpark.java:
```diff
@@ -0,0 +1,116 @@
+        }
+        JavaRDD<Variant> bqsrKnownVariants = variantsSparkSource.getParallelVariants(knownVariants.get(0));
+
+        GCSOptions options = BucketUtils.getAuthenticatedGCSOptions(apiKey);
```
What if we're not using GCS?  This will give an NPE.

droazen replied:
This version of the tool requires a GCS API key to access the reference data. The existing code will throw an `IllegalArgumentException` if no API key is specified, but we should make this a bit nicer with an explicit check that throws a `UserException` with a clear message.

tomwhite replied:
It should be possible to read reference data from the file system too (a FASTA file from the local filesystem or HDFS), see https://github.com/broadinstitute/hellbender/blob/master/src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java#L37.

This simplest fix here would be to check if `apiKey` is null, and if it is set `options` to null.

droazen replied:
Fixed

laserson commented on src/main/java/org/broadinstitute/hellbender/tools/spark/BaseRecalibratorSpark.java:
```diff
@@ -0,0 +1,116 @@
+
+        // TODO: Look into broadcasting the reference to all of the workers. This would make AddContextDataToReadSpark
+        // TODO: and ApplyBQSRStub simpler (#855).
+        JavaPairRDD<GATKRead, ReadContextData> rddReadContext = AddContextDataToReadSpark.add(initialReads, referenceDataflowSource, bqsrKnownVariants);
```
Need to filter out unmapped reads first?

droazen replied:
Indeed we do -- will add a filtering step.

droazen replied:
Decided to modify `AddContextDataToReadSpark` itself to filter out unmapped reads, since it can't currently handle them.

droazen commented:
@davidadamsphd Sure, here is a quick guide to the code:

-`BaseRecalibratorSpark` is the standalone BQSR tool, and calls into the `BaseRecalibratorSparkFn` (which is also called from `ReadsPipelineSpark`)

-`ApplyBQSRSpark` is the standalone ApplyBQSR tool, and calls into the `ApplyBQSRSparkFn` (also called from `ReadsPipelineSpark`)

-Integration tests for the above are in `BaseRecalibratorSparkIntegrationTest` and `ApplyBQSRSparkIntegrationTest`

-Almost all other changes in the branch are related to the BQSR engine refactoring, which I summarize below:

* We pulled out the guts of the walker `BaseRecalibrator` tool, combined it with all of the code from the former `RecalibrationEngine` class (now deleted) to make a new `BaseRecalibrationEngine` class under `utils/recalibration`.
 
* We stripped out all copies of the code in `BaseRecalibrationEngine` from the walker, dataflow, and spark versions of BQSR, and modified them to call into `BaseRecalibrationEngine`.

* We moved all auxiliary classes needed by the  `BaseRecalibrationEngine` (eg., the covariates, etc.) into `utils/recalibration`.

* We refactored the argument collections. Now there is a single shared `RecalibrationArgumentCollection` that contains **only** the parameters for the `BaseRecalibrationEngine` itself, and this argument collection is exposed by all 3 versions of the tool. Input/output arguments have been removed from this argument collection and put into the individual implementations of BQSR, since they vary between the walker, dataflow, and spark versions of the tool. This eliminates awkward problems such as having both a `knownSites` argument AND a `BQSRKnownVariants` exposed at the same time, with only 1 of them usable for a given version of a tool. The dataflow-only `BaseRecalibrationArgumentCollection` has been deleted completely as no longer needed.

* We tweaked the names of some tool arguments to enforce consistency between the 3 versions of the tool as well as the rest of hellbender (eg., output arg for BQSR is now a more standard `-O`)

jean-philippe-martin commented:
I haven't looked yet but this seems like a good idea. You should also take calculateSNPInsertDelete from https://github.com/broadinstitute/hellbender/blob/jp_perf_br/src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java#L212 since it's marginally faster than the calculateIs* combo.

droazen commented:
Since @davidadamsphd is out today, would either @tomwhite or @laserson be available to review JUST the spark part of this branch (specifically, just the classes `BaseRecalibratorSpark`, `BaseRecalibratorSparkFn`, `ApplyBQSRSpark`, `ApplyBQSRSparkFn`, `ReadsPipelineSpark`, `BaseRecalibratorSparkIntegrationTest`, and `ApplyBQSRSparkIntegrationTest`)? 

The rest of the branch is just a BQSR engine refactoring that will be reviewed by @jean-philippe-martin as soon as we do another rebase on top of master (we'll ping him as soon as that part of the branch is ready for review).

tomwhite commented:
@droazen I'll take a look.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/BaseRecalibratorSparkFn.java:
```diff
@@ -0,0 +1,63 @@
+
+                bqsr.processRead(readWithData._1(), refDS, variants);
+            }
+            return new ArrayList<>(Arrays.asList(bqsr.getRecalibrationTables()));
```
No need to wrap in `ArrayList`.

droazen replied:
Sadly, we do need to wrap this, as we can't get the special immutable `ArrayList` returned by `Arrays.asList()` to serialize despite some effort.

droazen replied:
(we tried both `ArraysAsListSerializer` and `UnmodifiableCollectionsSerializer`)

lbergelson replied:
We tried registering `de.javakaffee.kryoserializers.ArraysAsListSerializer`, but it still wanted to use `com.twitter.chill.java.ArraysAsListSerializer` despite the explicit registration.  This fails when it's Arrays.asList() of some primitive type.

tomwhite replied:
Ah OK. Maybe add a comment explaining this.

lbergelson replied:
@tomwhite Any idea WHY the com.twitter one would be used after explicitly registering the de.javakafee serializer?  Or how we can fix it?

tomwhite replied:
@lbergelson There's a comment in the Spark code saying that the Chill classes are registered after the user classes so that the user classes take effect: https://github.com/apache/spark/blob/v1.4.1/core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala#L119-L121

However, in Kryo the last registration wins: https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/Kryo.java#L410-L423

So it looks like you can't override Chill serializers, which looks like a bug to me.

BTW you can debug Kryo by calling `com.esotericsoftware.minlog.Log.TRACE()`.

droazen replied:
Added an explanatory comment here

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/ApplyBQSRSpark.java:
```diff
@@ -0,0 +1,81 @@
+public final class ApplyBQSRSpark extends SparkCommandLineProgram{
+    private static final long serialVersionUID = 0l;
+
+    @Argument(doc = "uri for the input bam, either a local file path or a gs:// bucket path",
```
Nit: or HDFS.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/ApplyBQSRSpark.java:
```diff
@@ -0,0 +1,81 @@
+        final JavaRDD<GATKRead> recalibratedReads = ApplyBQSRSparkFn.apply(initialReads, recalibrationReportBroadCast, readsHeader, applyBQSRArgs);
+
+        try {
+            ReadsSparkSink.writeReads(ctx, output, recalibratedReads, readsHeader, ReadsWriteFormat.SINGLE);
```
This should use the 'parallelism' code from https://github.com/broadinstitute/hellbender/blob/master/src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortBamSpark.java#L54.

droazen replied:
Added a `shardedOutput` argument to the tool that allows you to use `ReadsWriteFormat.SHARDED` here (off by default).

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java:
```diff
@@ -77,31 +81,27 @@ protected void runPipeline(final JavaSparkContext ctx) {
+        final JavaRDD<GATKRead> finalReads = ApplyBQSRSparkFn.apply(markedReads, reportBroadcast, readsHeader, new ApplyBQSRArgumentCollection());
 
         try {
             ReadsSparkSink.writeReads(ctx, output, finalReads, readsHeader, ReadsWriteFormat.SHARDED);
```
Similarly (parallelism).

droazen replied:
Added an option here as well to select between `ReadsWriteFormat.SINGLE` and `ReadsWriteFormat.SHARDED`

tomwhite commented:
I reviewed the Spark-specific parts and they look good to me. It would be nice to add some tests for using a local reference, but that doesn't need to hold this up. +1

droazen commented:
Thanks for the review @tomwhite!

droazen commented:
Ok, @jean-philippe-martin, I've finished rebasing on top of your work, and we're now ready for you to review the NON-spark parts of this branch.

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java:
```diff
@@ -116,13 +89,20 @@ public void processElement(ProcessContext c) throws Exception {
+        if ( read instanceof SAMRecordToGATKReadAdapter ) {
+            // we ship those without the header, put it back.
+            SAMRecordToGATKReadAdapter readAdapter = (SAMRecordToGATKReadAdapter)read;
+            readAdapter.setHeader(header);
```
I suspect this is unnecessary, as the tests were passing before and we weren't doing that. Perhaps the BaseRecal doesn't call any of the functions that needs a header?

At any rate, we need a utility function for that. No point in repeating this same code all over the place.

droazen replied:
Ok, pulled out a common utility here.

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java:
```diff
@@ -154,414 +133,16 @@ private void init(ProcessContext c) {
-            }
+    private void saveReport( File output, RecalibrationTables rt, QuantizationInfo quantizationInfo, StandardCovariateList finalRequestedCovariates ) throws FileNotFoundException {
+        try ( PrintStream reportStream = new PrintStream(output) ) {
+            RecalUtils.outputRecalibrationReport(reportStream, recalArgs, quantizationInfo, rt, finalRequestedCovariates, recalArgs.SORT_BY_ALL_COLUMNS);
```
Thank you for removing the dependency on the global variable.

droazen replied:
You're welcome!

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorOptimizedFn.java:
```diff
@@ -164,411 +138,16 @@ private void init(ProcessContext c) {
-        // luckily we don't use BRAC.RAC.RECAL_TABLE without first coming through here, so this is fine.
-        BRAC.RAC.RECAL_TABLE = new PrintStream(output);
-        RecalUtils.outputRecalibrationReport(BRAC.RAC, quantizationInfo, rt, finalRequestedCovariates, BRAC.RAC.SORT_BY_ALL_COLUMNS);
+        try ( PrintStream reportStream = new PrintStream(output) ) {
```
You know, if saveReport now deals with output streams, we should make that the interface (instead of a file) so that we can write to GCS again. Like this:

```java
saveReport(OutputStream output, ...) {
try (PrintStream reportStream = new PrintStream(output) { 
```

and we'd call it like this: `BaseRecalibratorDataflow.saveReport(BucketUtils.createFile(output, pipelineOptions), ...`

Alternatively we can pass the url (String) to saveReport, and it can call createFile internally.

droazen replied:
That would be a good feature, but is beyond the scope of this branch. Feel free to do a separate PR to add it after this one is merged.

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/tools/recalibration/BQSRGatherer.java:
```diff
@@ -1,84 +0,0 @@
-/**
- * Merges BQSR Recaliration tables.
- */
-public final class BQSRGatherer extends Gatherer {
```
I'm not sure that we can remove this class. It's used by the non-Dataflow version of BaseRecalibrator. It's done in a magic way, via annotation, but removing this may break the non-cloud version.

So removing this is a big step. If that's what we mean to do then it should at a minimum be mentioned in the PR description.

droazen replied:
It's not actually being used -- the `@Gather` annotation is only used by the GATK3 job runner Queue, which we don't support (yet). If we want to support Queue in the future we can restore this, but we'll need to make a bunch of other changes to our codebase like using `@Input` and `@Output` instead of `@Argument` for tool inputs and outputs.

jean-philippe-martin commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/BucketUtils.java:
```diff
@@ -173,4 +203,15 @@ public static boolean fileExists(String path, PipelineOptions popts) {
+     * Create a new GCS options with a given api key.
+     * @param apiKey a valid GCS api key
+     */
+    public static GCSOptions getAuthenticatedGCSOptions(String apiKey) {
```
This is dangerous. Authentication can be done via either an apiKey, or a client_secrets.json. We are currently limited to only testing the former on Travis, which means we have to be extra careful to make sure the latter keeps on working.

That's why I would much rather see the authentication information saved into the pipeline options and then used from there. The code there understands both api keys and client_secrets. We should not introduce paths that only understand one but not the other.

droazen replied:
This was just a convenience method added because spark currently only supports the apiKey, but your point about not introducing code paths that understand one but not the other is fair -- I've removed this.

droazen replied:
Ended up moving this to `SparkCommandLineProgram` to keep it spark-only.

jean-philippe-martin commented:
Done with my review. Thanks for doing this much-needed refactor! The BaseRecal stuff looks sound, but I have some other feedback that we should discuss/address.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/dataflow/BucketUtils.java:
```diff
@@ -72,6 +91,17 @@ public static InputStream openFile(String path, PipelineOptions popts) {
+     * @param apiKey a GCS api key for accessing gcs paths, may be null if you will never open a gcs file
+     * @return an InputStream that reads from the specified file.
+     */
+    public static InputStream openFile(String path, String apiKey) {
```
Removed

droazen commented:
All code review comments have been addressed, but I need to wait for https://github.com/broadinstitute/hellbender/pull/922 to be merged and rebase on top of it before tests will pass in this branch, as the spark tests are causing us to hit the travis log file limit.

laserson commented:
@droazen: merged #922 and rebased this branch on master.  I also pushed the `dr_lb_spark_bqsr_bak` branch which is pre-rebase in case there's a problem.  Merge if Travis passes?

droazen commented:
Made a final change or two, will merge once the latest round of tests pass.
